src/jira/authenticate.ts

```diff
@@ -1,28 +1,41 @@
-import { Installation } from "../models";
-import { hasValidJwt } from "./util/jwt";
-import { NextFunction, Request, Response } from "express";
+import {TokenType, verifyAsymmetricJwtTokenMiddleware, verifySymmetricJwtTokenMiddleware} from "./util/jwt";
+import {NextFunction, Request, Response} from "express";
+import {booleanFlag, BooleanFlags} from "../config/feature-flags";
 
-export default async (
+const authenticateJiraEventMiddleware = async (
 	req: Request,
 	res: Response,
-	next: NextFunction
+	next: NextFunction,
+	asymmetric: boolean
 ): Promise<void> => {
-	const installation = await Installation.getForClientKey(req.body.clientKey);
-	if (!installation) {
-		res.status(404).json({});
-		return;
+
+	if( asymmetric ) {
+		await verifyAsymmetricJwtTokenMiddleware(req, res, next);
+	} else {
+		verifySymmetricJwtTokenMiddleware(res.locals.installation.sharedSecret, TokenType.normal, req, res, next)
+	}
+
+};
+
+export const authenticateJiraEvent = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+	await authenticateJiraEventMiddleware(req, res, next, false)
+}
+
+export const authenticateUninstallCallback = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+
+	if (await booleanFlag(BooleanFlags.USE_JWT_SIGNED_INSTALL_CALLBACKS, false) ) {
+		await authenticateJiraEventMiddleware(req, res, next, true)
+	} else {
+		await authenticateJiraEventMiddleware(req, res, next, false)
 	}
+}
 
-	const { jiraHost, sharedSecret, clientKey } = installation;
 
-	req.addLogFields({
```

Comment:
I've extracted fetching installation and adding it to the local context to a separate middleware function. See `src/jira/extract-installation.ts`